### PR TITLE
feat: Implement Official Exchange Rates Retrieval

### PR DESCRIPTION
**Pull request by funwie**

### CODE REVIEW

**Comment by funwie** [comment]
Thank you for taking the initiate to implement the official exchange rates. You've made impressive progress. However, there are few improvements we need to make for the API to be production ready.

I will discuss general areas of improvement below and leave comments in the code too. I am available to work on improvements together if necessary.

### Learn from existing code and colleagues
- Our existing code repositories are a great place to learn how we write code.
- Reach out to colleagues to learn why we write code the way we do or to help as necessary.
- It's highly important that we keep all our code repos consistent to ease transition between them.
- If you think there is a better way to do things, share with the team and let's learn about it together.

### Break up requirements into smaller tasks
We are setting up a new solution, and implementing requirements in the same task/branch. Having different changes in a single task/branch lengthens the review and delays the changes.
- Scaffold the solution, projects, README in a separate task
- Implement the exchange rate requirements in the next task

### Ensure the Implementation meets the Requirements
Always seek clarification to understand the requirements well before implementation
The requirements state that `We need an HTTP service which accepts a source currency and target currencies and returns the latest rates. ` However, the current implementation only accepts target currencies. Source currency is optional not 
Add source currency to the request  and make it optional. Default to `Czech koruna (CKZ)` when a value is not provided

### Consider how more source currencies will be added
The requirement states that `A source currency may be optional for now, within the first iteration we plan to support only CZK as a source currency`. This implies that we plan to support other source currencies in the future. 

Use the factory design pattern to make code easier to extend for other source currencies
Design pattern is a solution to a common problem faced in software development. Learn more at [refactoring.guru](https://refactoring.guru) 

### Always Validate Input
We are not validating the target currencies from the API Users. 
It is a good practise to always validate input and return appropriate error response to User if invalid.
Validation also prevents unnecessary processing of invalid requests.
Few examples to validate:
- What if a source or target currency has more than 3 letters? `USDD`
- What if a source or target currency has 3 characters? `US1`
- What if zero target currencies are provided?

### Follow Software Principles like SOLID, KISS, and more
I understand that having all the code in a single `program.cs` file keeps things simple. However, many ideas in a single file creates complexity and hiders future maintenance or extension of the software.

Software principles guide us to build highly maintainable and extensible software. Following principles will allow your future self and other engineers to easily understand the codebase, to make new changes comfortably. 

Our other code repos will show the principles we apply and you can learn more in links below
- [A short intro to software principles](https://thebulbafrica.medium.com/5-software-development-principles-that-should-be-embraced-daily-75a6f6cc00ff)
- [SOLID principles ](https://stackoverflow.blog/2021/11/01/why-solid-principles-are-still-the-foundation-for-modern-software-architecture/)

### Caching
A cache stores data temporally so that future requests for that data are served up faster than is possible by accessing the data from the primary source.

We retrieve daily rates from Czech National Bank (CNB) API every time a request is sent to our API. Given that the rates do not change throughout the day, we can retrieve once a day, and cache for the entire day. This prevents us from calling the CNB rates endpoint multiple times a day for the same data. 

Implement in-memory cache to store all the rates returned from the CNB API. Store all the rates so that if the target currencies change, cached response can still produce the target rates.

[Learn how to Cache in-memory in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-8.0)

### Testing
No part of the code is tested.
Testing is one part of building quality software. Testing validates that:
- Code satisfies the requirements with correct logic
- Errors are handled properly
- Edge cases are considered and handled properly
- Future changes do not break existing code

Add following tests
- Unit Test for units of logic in our code and their edge cases
- Integration Test for the integration with external CNB API

Learn more about testing using links:
[Unit testing C# in .NET using dotnet test and xUnit](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-with-dotnet-test)
[Unit Testing in ASP.NET Core Web API](https://code-maze.com/unit-testing-aspnetcore-web-api/)
[Integration tests in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-8.0)

### Observability
This involves adding features to our code to:
- Measure the API's state to know if it is running as expected or unexpected
- Produce information that will help us investigate incidents in the API

If this API is deployed and something goes wrong, we will not know and it will be impossible to investigate what's happening. We can add logging for now to improve the API's observability
- Log errors that happen in the API
- Log the request and response. This can be implemented using a middleware

Lear more at:
[ASP.NET Core Middleware](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-8.0)
[Logging in .NET Core and ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-8.0)
[What is observability?](https://newrelic.com/blog/best-practices/what-is-observability)


I have provided links to learn more about concepts but always start by learning from our existing code repos, and from colleagues.

File: BackendTaskPilot/BackendTaskPilot.csproj [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+    <PropertyGroup>
+        <TargetFramework>net7.0</TargetFramework>
```
Comment (funwie) [inline BackendTaskPilot/BackendTaskPilot.csproj:4]: If our other services are on the latest `net8.0` version, then this project should also be on `net8.0`

File: BackendTaskPilot.sln [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "BackendTaskPilot", "BackendTaskPilot\BackendTaskPilot.csproj", "{85AE9448-BC75-4412-8D41-DA6C52B676D9}"
```
Comment (funwie) [inline BackendTaskPilot.sln:3]: #### Project Structure, Naming, and Separation
This implementation has everything in a single project. It's good practise to separate different concerns into their own projects. 

- Separate logic into different projects so it's easier to maintain and extend
- Place all projects in a `src` folder and all test projects in a `test` folder. 
- Give the solution and projects names that describe the domain. 
- Follow same structure in our other repositories or below structure if no repos exist
- Learn more using links below
  - [Common web application architectures](https://learn.microsoft.com/en-us/dotnet/architecture/modern-web-apps-azure/common-web-application-architectures)
  - [C# at Google Style Guide](https://google.github.io/styleguide/csharp-style.html)
  - [eShop](https://github.com/dotnet/eShop) 

Example:

ExchangeRate - solution name
 - src
   - ExchangeRate.Api
   - ExchangeRate.Application
 - test
   - ExchangeRate.Api.UnitTest 
   - ExchangeRate.Api.IntegrationTest 
   - ExchangeRate.Application.UnitTest

File: BackendTaskPilot/Program.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+ExchangeRateService service = new ExchangeRateService();
+
+// REST API to get exchange rates based on input currencies
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:9]: Comment is unnecessary because the code is self-explanatory

File: BackendTaskPilot/Program.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+public struct CnbRatesResponse
+{
+    [JsonPropertyName("rates")] public List<Rate> Rates { get; set; }
+}
+
+public struct Rate
+{
+    [JsonPropertyName("currencyCode")] public string CurrencyCode { get; set; }
+    [JsonPropertyName("rate")] public float RateValue { get; set; }
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:22]: - Place `JsonPropertyName` annotation on separate line to improve readability
- Rename RateValue to Rate to reflect the property name in the CNB response
- Change the property type to decimal to handle precision of the value properly
- Make property `init` only to prevent mutation after it's created
```suggestion
    [JsonPropertyName("rate")]
    public decimal Rate { get; init; }
```

File: BackendTaskPilot/Program.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+internal class ExchangeRateService
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:25]: The ExchangeRateService class can be improved by following SOLID principles
- Implement the IExchangeRateService interface
- place IExchangeRateService interface and ExchangeRateService in separate files with same name respectively
- Both can be placed in separate `ExchangeRate.Application` project to separate the business code from http api code

Use the factory design pattern to easily extend our code to handle other source currencies in the future. 
Implement two interfaces similar to below

Each source currency will implement this interface to retrieve the rates from the official external api
```csharp
public interface IExchangeRateProvider
{
    public string CurrencyCode { get; init; }
    Task<IEnumerable<ExchangeRate>> GetDailyExchangeRatesAsync(DateOnly day, CancellationToken cancellationToken);
}
```

// This interface will use the factory design pattern to find the provider for a given source currency
```csharp
public interface IExchangeRateProviderFactory
{
    IExchangeRateProvider Get(string currencyCode);
}
```

File: BackendTaskPilot/Program.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+{
+    private const string MainCurrenciesUrl = "https://api.cnb.cz/cnbapi/exrates/daily";
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:27]: It's good practise to place API settings in `appsettings.json`. Settings that
- Can change depending on the environment 
- Can be changed without changing the code

Make this URL Configurable 
- Add CnbBaseUrl with value `https://api.cnb.cz/` to appsettings.json file and inject it using IOptions
- Use the `/cnbapi/exrates/daily` endpoint as request URI when retrieving daily rates. Other endpoints can be used based on the type of data to retrieve from the same `https://api.cnb.cz/` API

See [Configuration in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-8.0) for more information

File: BackendTaskPilot/Program.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    public IEnumerable<Rate> GetExchangeRates(List<string> currencies)
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:29]: This method should be replaced with implementation of the `IExchangeRateService` interface

File: BackendTaskPilot/appsettings.json [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{
+  "Logging": {
+    "LogLevel": {
+      "Default": "Information",
+      "Microsoft.AspNetCore": "Warning"
+    }
+  },
+  "AllowedHosts": "*"
```
Comment (funwie) [inline BackendTaskPilot/appsettings.json:8]: Add CNB base url configuration
```suggestion
  "AllowedHosts": "*",
  "ExternalServices": {
    "CnbBaseUrl": "https://api.cnb.cz/"
  }
```

See [Configuration in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-8.0)  to use this config

File: BackendTaskPilot/Program.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+ExchangeRateService service = new ExchangeRateService();
+
+// REST API to get exchange rates based on input currencies
+app.MapPost("/", service.GetExchangeRates);
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:10]: #### REST API Design
This POST endpoint is at the API root which does not follow good REST API design. 
The API should have a  `exchange-rate` resource and a GET endpoint to retrieve the rates. 
`GET /exchange-rate` should accept the source currency and target currencies in query strings. 

Ideally, the API design should have been done separately, and before this implementation so that resource names, endpoints, request & response contracts are defined
[Learn more about REST APIs](https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/)

Following separation of concerns principle, 
- Use a request handler instead of passing the `service.GetExchangeRates` directly.
- `IExchangeRateService` should be injected into the handler to get the exchange rates
- The handler should be responsible for extracting source currency and target currencies from the request query strings, validate both inputs, and pass inputs to the service if valid
- The handler should return success or error responses depending on the result of get rates call

We will have code similar to below
```suggestion
app.MapGet("/exchange-rates", async (
    [FromQuery] string? sourceCurrency,
    [FromQuery] string[] targetCurrencies,
    [FromServices] IExchangeRateService rateService,
    CancellationToken cancellationToken = default) =>
{
    // validate sourceCurrency and targetCurrencies
    // Use ExchangeRateService to get daily rates
    // Return success response with rates
    // OR return error response
});
```

File: BackendTaskPilot/Program.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+ExchangeRateService service = new ExchangeRateService();
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:7]: This service can be improved following SOLID principles. 
- Code to an `IExchangeRateService` interface
- Instead of instantiating the service manually, use dependency injection (DI) to inject the service for use in other areas of the application
- The interface will also enable us to mock the service for unit testing

See  [Dependency injection in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-8.0)  for more

Example interface
```csharp
public interface IExchangeRateService
{
    Task<IEnumerable<ExchangeRate>> GetDailyRatesAsync(string? sourceCurrency, string[] targetCurrencies, CancellationToken cancellationToken = default);
}
```

File: BackendTaskPilot/Program.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+app.MapPost("/", service.GetExchangeRates);
+
+app.Run();
+
+public struct CnbRatesResponse
+{
+    [JsonPropertyName("rates")] public List<Rate> Rates { get; set; }
+}
+
+public struct Rate
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:19]: This `Rate` type is shared between CNB response contract and our exchange rate api response contract. A change made for one could implicitly affect the other.
- Let's create a separate `ExchangeRate` type for our exchange rate api response contract
- Let's rename `Rate` class to `CnbRate` to reflect its origin
- Both should be of type `class` or `record` and not a `struct`. See [Choosing Between Class and Struct](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct)

```suggestion
public class CnbRate
```

File: BackendTaskPilot/Program.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+{
+    [JsonPropertyName("currencyCode")] public string CurrencyCode { get; set; }
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:21]: - Place `JsonPropertyName` annotation on separate line to improve readability 
- Make property `init` only to prevent mutation after it's created
```suggestion
    [JsonPropertyName("currencyCode")] 
    public string CurrencyCode { get; init; }
```

File: BackendTaskPilot/Program.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    [JsonPropertyName("rate")] public float RateValue { get; set; }
+}
+
+internal class ExchangeRateService
+{
+    private const string MainCurrenciesUrl = "https://api.cnb.cz/cnbapi/exrates/daily";
+
+    public IEnumerable<Rate> GetExchangeRates(List<string> currencies)
+    {
+        HttpClient client = new HttpClient();
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:31]: #### Instantiating a new HTTP client on every request is not performant
- HTTP client is a resource intensive object and creating many instances will likely end in socket exhaustion and resource starvation. 
- Use the Typed Client that is a performant way to manage HTTP clients in .Net. 
- The Typed Client will implement the `IExchangeRateProvider` interface to retrieve `CZK` source currency rates.
[Learn how to use Type Clients in ASP.net core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-8.0#typed-clients)

File: BackendTaskPilot/Program.cs [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+			
+        var data = client.GetStringAsync(MainCurrenciesUrl).Result;
+        var response = JsonSerializer.Deserialize<CnbRatesResponse>(data);
+
+        return response.Rates.Select(rate => rate).Where(rate => currencies.Contains(rate.CurrencyCode));
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:36]: - Before returning the results. Store the rates from provider in the in-memory cache to expire at end of day. 
- Cache can use a combination of source currency + day as key
```suggestion
        return exchangeRates.Where(r => targetCurrencies.Contains(r.CurrencyCode));
```

Not relevant in suggested code but it's better to place `Where` before the `Select` so that we filter the results first, and then map only rates that we need.

File: BackendTaskPilot/Program.cs [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+internal class ExchangeRateService
+{
+    private const string MainCurrenciesUrl = "https://api.cnb.cz/cnbapi/exrates/daily";
+
+    public IEnumerable<Rate> GetExchangeRates(List<string> currencies)
+    {
+        HttpClient client = new HttpClient();
+			
+        var data = client.GetStringAsync(MainCurrenciesUrl).Result;
+        var response = JsonSerializer.Deserialize<CnbRatesResponse>(data);
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:34]: A critical takeaway from here is to avoid using `.Result` and instead `await` the task
Learn why at [Avoiding simple mistakes in async await](https://www.anthonysteele.co.uk/AsyncBasicMistakes.html)

Check the in-memory cache first, and return rates if found, else get rates from provider

```csharp

```
```suggestion
// default the source currency if not provided
sourceCurrency ??= CurrencyCodes.CZK;

// Create the day to pass to external API. This makes it explicit that we are retrieving for today
var day = DateOnly.FromDateTime(DateTime.UtcNow);

// Handle errors and exceptions in lines below and return appropriate results

// get the exchange rate provider for source currency
var provider = await exchangeRateProviderFactory.Get(sourceCurrency);

// Get the exchange rates from the provider
var exchangeRates = provider.GetDailyExchangeRatesAsync(day, cancellationToken);
```

File: BackendTaskPilot/Program.cs [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+ExchangeRateService service = new ExchangeRateService();
+
+// REST API to get exchange rates based on input currencies
+app.MapPost("/", service.GetExchangeRates);
+
+app.Run();
+
+public struct CnbRatesResponse
+{
+    [JsonPropertyName("rates")] public List<Rate> Rates { get; set; }
```
Comment (funwie) [inline BackendTaskPilot/Program.cs:16]: - Place `JsonPropertyName` annotation on separate line to improve readability 
- Make property `init` only to prevent mutation after it's created
- Use the IReadOnlyCollection<T> to make property readonly
```suggestion
    [JsonPropertyName("rates")] 
    public IReadOnlyCollection<CnbRate> Rates { get; init; }
```
